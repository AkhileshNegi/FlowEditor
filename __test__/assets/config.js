const config = {
    flow: 'boring'
};

module.exports =
    process.env.NODE_ENV === 'production'
        ? Object.assign({}, config, {
              localStorage: true,
              endpoints: {
                  flows: 'flows',
                  groups: 'groups',
                  contacts: 'contacts',
                  fields: 'fields',
                  labels: 'labels',
                  activity: '',
                  engine: ''
              }
          })
        : Object.assign({}, config, {
              localStorage: true,
              endpoints: {
                  flows: '/assets/flows',
                  groups: '/assets/groups',
                  contacts: '/assets/contacts',
                  fields: '/assets/fields',
                  labels: '/assets/labels',
                  activity: '',
                  engine: ''
              }
          });
