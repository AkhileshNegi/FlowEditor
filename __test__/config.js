const config = {
    flow: 'a4f64f1b-85bc-477e-b706-de313a022979'
};

module.exports =
    process.env.NODE_ENV === 'production'
        ? Object.assign({}, config, {
              localStorage: true,
              endpoints: {
                  flows: 'flows',
                  groups: 'groups',
                  recipients: 'recipients',
                  fields: 'fields',
                  labels: 'labels',
                  languages: 'languages',
                  activity: '',
                  simulateStart: '',
                  simulateResume: ''
              }
          })
        : Object.assign({}, config, {
              localStorage: true,
              endpoints: {
                  flows: '/assets/flows.json',
                  groups: '/assets/groups.json',
                  fields: '/assets/fields.json',
                  recipients: '/assets/recipients.json',
                  labels: '/assets/labels.json',
                  languages: '/assets/languages.json',
                  activity: '',
                  simulateStart: '',
                  simulateResume: ''
              }
          });
