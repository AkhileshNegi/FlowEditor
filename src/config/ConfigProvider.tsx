import * as React from 'react';

import { Endpoints, FlowEditorConfig } from '../flowTypes';
import AssetService from '../services/AssetService';

export const fakePropType: any = (): any => null;
fakePropType.isRequired = (): any => null;

interface ConfigProviderProps {
    config: FlowEditorConfig;
    assetService: AssetService;
    children: any;
}

export interface ConfigProviderContext {
    assetService: AssetService;
    endpoints: Endpoints;
    flow: string;
}

// ----------------------------------------------------------------------------------------------

export const SINGLE_CHILD_ERROR = 'ConfigProvider expects only one child component.';
export const VALID_CHILD_ERROR =
    'ConfigProvider expects a valid React element: https://reactjs.org/docs/react-api.html#isvalidelement';

export default class ConfigProvider extends React.Component<ConfigProviderProps> {
    public static childContextTypes = {
        assetService: fakePropType,
        endpoints: fakePropType,
        flow: fakePropType
    };

    constructor(props: ConfigProviderProps) {
        super(props);

        if (React.Children.count(props.children) > 1) {
            throw new Error(SINGLE_CHILD_ERROR);
        } else if (!React.isValidElement(props.children)) {
            throw new Error(VALID_CHILD_ERROR);
        }
    }

    public getChildContext(): ConfigProviderContext {
        return {
            assetService: this.props.config.assetService,
            endpoints: this.props.config.endpoints,
            flow: this.props.config.flow
        };
    }

    public render(): JSX.Element {
        return React.Children.only(this.props.children);
    }
}
