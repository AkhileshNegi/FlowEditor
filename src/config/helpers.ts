import { Operator, Type } from '~/config';
import { FlowTypes, FlowTypeVisibility } from '~/config/interfaces';

export const filterOperators = (operators: Operator[], flowType: FlowTypes): Operator[] => {
    // return filterVisibility(operators, flowType);
    return operators;
};

export const filterTypeConfigs = (typeConfigs: Type[], flowType: FlowTypes): Type[] => {
    return filterVisibility(typeConfigs, flowType);
};

const filterVisibility = (items: FlowTypeVisibility[], flowType: FlowTypes): any[] => {
    return items.filter((item: FlowTypeVisibility) => {
        if (item.visibility === undefined) {
            return true;
        }

        return item.visibility.findIndex((ft: FlowTypes) => ft === flowType) > -1;
    });
};
