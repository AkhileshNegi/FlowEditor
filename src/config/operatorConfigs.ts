import { HIDDEN, Operator, OperatorMap, Operators, TEXT_TYPES } from '~/config/interfaces';

export const operatorConfigList: Operator[] = [
    {
        type: Operators.has_any_word,
        verboseName: 'has any of the words',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_all_words,
        verboseName: 'has all of the words',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_phrase,
        verboseName: 'has the phrase',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_only_phrase,
        verboseName: 'has only the phrase',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_beginning,
        verboseName: 'starts with',
        operands: 1
    },
    {
        type: Operators.has_text,
        verboseName: 'has some text',
        operands: 0,
        categoryName: 'Has Text',
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_number,
        verboseName: 'has a number',
        operands: 0,
        categoryName: 'Has Number'
    },
    {
        type: Operators.has_number_between,
        verboseName: 'has a number between',
        operands: 2
    },
    {
        type: Operators.has_number_lt,
        verboseName: 'has a number below',
        operands: 1
    },
    {
        type: Operators.has_number_lte,
        verboseName: 'has a number at or below',
        operands: 1
    },
    {
        type: Operators.has_number_eq,
        verboseName: 'has a number equal to',
        operands: 1
    },
    {
        type: Operators.has_number_gte,
        verboseName: 'has a number at or above',
        operands: 1
    },
    {
        type: Operators.has_number_gt,
        verboseName: 'has a number above',
        operands: 1
    },
    {
        type: Operators.has_date,
        verboseName: 'has a date',
        operands: 0,
        categoryName: 'Has Date',
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_date_lt,
        verboseName: 'has a date before',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_date_eq,
        verboseName: 'has a date equal to',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_date_gt,
        verboseName: 'has a date after',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_run_status,
        verboseName: 'has a run status of',
        operands: 1,
        visibility: HIDDEN
    },
    {
        type: Operators.has_group,
        verboseName: 'is in the group',
        operands: 1,
        visibility: HIDDEN
    },
    {
        type: Operators.has_phone,
        verboseName: 'has a phone number',
        operands: 0,
        categoryName: 'Has Phone'
    },
    {
        type: Operators.has_email,
        verboseName: 'has an email',
        operands: 0,
        categoryName: 'Has Email',
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_error,
        verboseName: 'has an error',
        operands: 0,
        categoryName: 'Has Error',
        visibility: HIDDEN
    },
    {
        type: Operators.has_value,
        verboseName: 'is not empty',
        operands: 0,
        categoryName: 'Not Empty'
    },
    {
        type: Operators.has_wait_timed_out,
        verboseName: null,
        operands: 0,
        visibility: HIDDEN
    },
    {
        type: Operators.has_pattern,
        verboseName: 'matches regex',
        operands: 1
    }
];

export const operatorConfigMap: OperatorMap = operatorConfigList.reduce(
    (map: OperatorMap, operatorConfig: Operator) => {
        map[operatorConfig.type] = operatorConfig;
        return map;
    },
    {}
);

/**
 * Shortcut for constant lookup of operator config in operator configs map
 * @param {string} type - The type of the operator config to return, e.g. 'send_msg'
 * @returns {Object} - The operator config found at operatorConfigs[type] or -1
 */
export const getOperatorConfig = (type: Operators): Operator => operatorConfigMap[type];
