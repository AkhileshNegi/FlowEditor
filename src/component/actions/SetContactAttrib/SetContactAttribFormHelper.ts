import { FormHelper, Types } from '../../../config/typeConfigs';
import { Action, SetContactAttribute, SetContactField, SetContactName } from '../../../flowTypes';
import {
    SetContactAttribFormState,
    SetContactFieldFormState,
    SetContactNameFormState
} from '../../../store/nodeEditor';
import { assetToField, fieldToAsset, propertyToAsset } from './helpers';

export class SetContactAttribFormHelper implements FormHelper {
    // passed an existing action or null
    public actionToState(
        action: SetContactAttribute,
        actionType: Types.set_contact_field | Types.set_contact_name
    ): SetContactAttribFormState {
        // if we have an existing contact attribute action, use it
        if (action) {
            switch (action.type) {
                case Types.set_contact_field:
                    return {
                        field: fieldToAsset(action.field),
                        value: action.value
                    } as SetContactFieldFormState;
                case Types.set_contact_name:
                    return {
                        name: propertyToAsset((action as SetContactName).type),
                        value: (action as SetContactName).name
                    } as SetContactNameFormState;
            }
        }

        // otherwise, provide initial form state from scratch
        switch (actionType) {
            case Types.set_contact_field:
                return {
                    field: fieldToAsset(),
                    value: ''
                } as SetContactFieldFormState;
            case Types.set_contact_name:
                return {
                    name: propertyToAsset(actionType),
                    value: ''
                } as SetContactNameFormState;
        }
    }

    public stateToAction(
        actionUUID: string,
        formState: SetContactAttribFormState,
        formType: Types
    ): SetContactAttribute {
        const action: Partial<Action> = {
            uuid: actionUUID,
            type: formType
        };

        switch (formType) {
            case Types.set_contact_field:
                return {
                    ...action,
                    field: assetToField((formState as SetContactFieldFormState).field),
                    value: (formState as SetContactFieldFormState).value
                } as SetContactField;
            case Types.set_contact_name:
                return {
                    ...action,
                    name: formState.value
                } as SetContactName;
        }
    }
}
