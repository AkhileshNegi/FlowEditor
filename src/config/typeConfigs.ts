import AddLabelsComp from '../component/actions/AddLabels/AddLabels';
import AddLabelsForm from '../component/actions/AddLabels/AddLabelsForm';
import CallWebhookComp from '../component/actions/CallWebhook/CallWebhook';
import AddGroupsForm from '../component/actions/ChangeGroups/AddGroupsForm';
import ChangeGroupsComp from '../component/actions/ChangeGroups/ChangeGroups';
import RemoveGroupsForm from '../component/actions/ChangeGroups/RemoveGroupsForm';
import MissingComp from '../component/actions/Missing/Missing';
import SendBroadcastComp from '../component/actions/SendBroadcast/SendBroadcast';
import SendBroadcastForm from '../component/actions/SendBroadcast/SendBroadcastForm';
import { SendBroadcastFormHelper } from '../component/actions/SendBroadcast/SendBroadcastFormHelper';
import SendEmailComp from '../component/actions/SendEmail/SendEmail';
import SendEmailForm from '../component/actions/SendEmail/SendEmailForm';
import SendMsgComp from '../component/actions/SendMsg/SendMsg';
import SendMsgForm from '../component/actions/SendMsg/SendMsgForm';
import SetContactAttrib from '../component/actions/SetContactAttrib/SetContactAttrib';
import SetContactAttribForm from '../component/actions/SetContactAttrib/SetContactAttribForm';
import { SetContactAttribFormHelper } from '../component/actions/SetContactAttrib/SetContactAttribFormHelper';
import SetRunResultComp from '../component/actions/SetRunResult/SetRunResult';
import SetRunResultForm from '../component/actions/SetRunResult/SetRunResultForm';
import StartFlowComp from '../component/actions/StartFlow/StartFlow';
import StartSessionComp from '../component/actions/StartSession/StartSession';
import StartSessionForm from '../component/actions/StartSession/StartSessionForm';
import { StartSessionFormHelper } from '../component/actions/StartSession/StartSessionFormHelper';
import GroupsRouter from '../component/routers/GroupsRouter';
import SubflowRouter from '../component/routers/SubflowRouter';
import SwitchRouter from '../component/routers/SwitchRouter';
import WebhookRouter from '../component/routers/WebhookRouter';
import { AnyAction, RouterTypes, UINodeTypes } from '../flowTypes';
import { NodeEditorForm } from '../store/nodeEditor';

/*
Old name	                New name	                Event(s) generated

add_urn	                    add_contact_urn	            contact_urn_added
add_to_group	            add_contact_groups	        contact_groups_added
remove_from_group	        remove_contact_groups	    contact_groups_removed
set_preferred_channel	    set_contact_channel	        contact_channel_changed
update_contact	            set_contact_name	        contact_name_changed
update_contact	            set_contact_language	    contact_language_changed
update_contact	            set_contact_timezone	    contact_timezone_changed
save_contact_field      	set_contact_field	        contact_field_changed
save_flow_result	        set_run_result	            run_result_changed
call_webhook	            call_webhook	            webhook_called
add_label	                add_input_labels	        input_labels_added
reply	                    send_msg	                msg_created
send_email	                send_email	                email_created / email_sent
send_msg	                send_broadcast	            broadcast_created
start_flow	                start_flow	                flow_triggered
start_session	            start_session	            session_triggered
*/

export const enum Types {
    add_contact_urn = 'add_contact_urn',
    add_contact_groups = 'add_contact_groups',
    add_input_labels = 'add_input_labels',
    remove_contact_groups = 'remove_contact_groups',
    set_contact_channel = 'set_contact_channel',
    set_contact_name = 'set_contact_name',
    set_contact_field = 'set_contact_field',
    set_run_result = 'set_run_result',
    call_webhook = 'call_webhook',
    send_msg = 'send_msg',
    send_email = 'send_email',
    send_broadcast = 'send_broadcast',
    start_flow = 'start_flow',
    start_session = 'start_session',
    split_by_expression = 'split_by_expression',
    split_by_groups = 'split_by_groups',
    wait_for_response = 'wait_for_response',

    missing = 'missing'
}

export enum Mode {
    EDITING = 0x1,
    TRANSLATING = 0x2,
    ALL = EDITING | TRANSLATING
}

export interface FormHelper {
    actionToState: (action: AnyAction, actionType?: Types) => NodeEditorForm;
    stateToAction: (actionUUID: string, formState: NodeEditorForm, formType?: Types) => AnyAction;
}

export interface Type {
    type: Types;
    name: string;
    description: string;
    allows(mode: Mode): boolean;
    component?: React.SFC<AnyAction>;
    form?: React.ComponentClass<any>;
    formHelper?: FormHelper;
    advanced?: Mode;
    aliases?: string[];
}

export interface TypeMap {
    [propName: string]: Type;
}

export type GetTypeConfig = (type: string) => Type;

export function allows(mode: Mode): boolean {
    return (this.advanced & mode) === mode;
}

const ContactAttribHelper = new SetContactAttribFormHelper();

export const typeConfigList: Type[] = [
    /** Actions */
    {
        type: Types.missing,
        name: 'Missing',
        description: ' ** Unsupported ** ',
        component: MissingComp,
        allows
    },
    {
        type: Types.send_msg,
        name: 'Send Message',
        description: 'Send the contact a message',
        form: SendMsgForm,
        component: SendMsgComp,
        advanced: Mode.EDITING,
        allows
    },
    {
        type: Types.send_broadcast,
        name: 'Send Broadcast',
        description: 'Send somebody else a message',
        form: SendBroadcastForm,
        formHelper: new SendBroadcastFormHelper(),
        component: SendBroadcastComp,
        allows
    },
    {
        type: Types.add_input_labels,
        name: 'Add Labels',
        description: 'Label the incoming message',
        form: AddLabelsForm,
        component: AddLabelsComp,
        allows
    },
    {
        type: Types.add_contact_groups,
        name: 'Add to Group',
        description: 'Add the contact to a group',
        form: AddGroupsForm,
        component: ChangeGroupsComp,
        allows
    },
    {
        type: Types.remove_contact_groups,
        name: 'Remove from Group',
        description: 'Remove the contact from a group',
        form: RemoveGroupsForm,
        component: ChangeGroupsComp,
        allows
    },
    {
        type: Types.set_contact_field,
        name: 'Update Contact',
        description: 'Update the contact',
        form: SetContactAttribForm,
        formHelper: ContactAttribHelper,
        component: SetContactAttrib,
        allows
    },
    {
        type: Types.set_contact_name,
        name: 'Update Contact',
        description: 'Update the contact',
        form: SetContactAttribForm,
        formHelper: ContactAttribHelper,
        component: SetContactAttrib,
        allows
    },
    {
        type: Types.send_email,
        name: 'Send Email',
        description: 'Send an email',
        form: SendEmailForm,
        component: SendEmailComp,
        allows
    },
    {
        type: Types.set_run_result,
        name: 'Save Flow Result',
        description: 'Save a result for this flow',
        form: SetRunResultForm,
        component: SetRunResultComp,
        allows
    },
    // {type: 'set_preferred_channel', name: 'Set Preferred Channel', description: 'Set their preferred channel', component: Missing},
    /** Hybrids */
    {
        type: Types.call_webhook,
        name: 'Call Webhook',
        description: 'Call a webook',
        form: WebhookRouter,
        component: CallWebhookComp,
        advanced: Mode.EDITING,
        aliases: [UINodeTypes.webhook],
        allows
    },
    {
        type: Types.start_flow,
        name: 'Start a Flow',
        description: 'Enter another flow',
        form: SubflowRouter,
        component: StartFlowComp,
        aliases: [UINodeTypes.subflow],
        allows
    },
    {
        type: Types.start_session,
        name: 'Start Somebody Else',
        description: 'Start somebody else in a flow',
        form: StartSessionForm,
        formHelper: new StartSessionFormHelper(),
        component: StartSessionComp,
        allows
    },

    /** Routers */
    {
        type: Types.split_by_expression,
        name: 'Split by Expression',
        description: 'Split by a custom expression',
        form: SwitchRouter,
        advanced: Mode.TRANSLATING,
        allows
    },
    {
        type: Types.split_by_groups,
        name: 'Split by Group Membership',
        description: 'Split by group membership',
        form: GroupsRouter,
        allows
    },
    {
        type: Types.wait_for_response,
        name: 'Wait for Response',
        description: 'Wait for the contact to respond',
        form: SwitchRouter,
        advanced: Mode.TRANSLATING,
        aliases: [RouterTypes.switch],
        allows
    }
    // {type: 'random', name: 'Random Split', description: 'Split them up randomly', form: RandomRouterForm}
];

export const typeConfigMap: TypeMap = typeConfigList.reduce((map: TypeMap, typeConfig: Type) => {
    map[typeConfig.type] = typeConfig;
    if (typeConfig.aliases) {
        typeConfig.aliases.forEach((alias: string) => (map[alias] = typeConfig));
    }
    return map;
}, {});

/**
 * Shortcut for constant lookup of type config in type configs map
 * @param {string} type - The type of the type config to return, e.g. 'send_msg'
 * @returns {Object} - The type config found at typeConfigs[type] or -1
 */
export const getTypeConfig = (type: Types | RouterTypes | UINodeTypes): Type => {
    let actionConfig = typeConfigMap[type];

    if (!actionConfig) {
        actionConfig = typeConfigMap.missing;
    }
    return actionConfig;
};
