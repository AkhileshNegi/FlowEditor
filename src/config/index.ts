import ConfigProvider, { ConfigProviderContext } from './ConfigProvider';
import { Type, Mode, typeConfigList, getTypeConfig } from './typeConfigs';
import { Operator, operatorConfigList, getOperatorConfig } from './operatorConfigs';

export {
    ConfigProviderContext,
    Type,
    Mode,
    Operator,
    typeConfigList,
    getTypeConfig,
    operatorConfigList,
    getOperatorConfig
};

export default ConfigProvider;
