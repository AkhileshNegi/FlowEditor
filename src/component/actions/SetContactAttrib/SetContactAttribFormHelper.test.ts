import { Types } from '../../../config/typeConfigs';
import {
    createSetContactFieldAction,
    createSetContactNameAction
} from '../../../testUtils/assetCreators';
import { SetContactAttribFormHelper } from './SetContactAttribFormHelper';

const formHelper = new SetContactAttribFormHelper();

const setContactFieldAction = createSetContactFieldAction();

const setContactNameAction = createSetContactNameAction();

const setContactFieldFormState = formHelper.actionToState(
    setContactFieldAction,
    Types.set_contact_field
);

const setContactNameFormState = formHelper.actionToState(
    setContactNameAction,
    Types.set_contact_name
);

describe('SetContactAttribFormHelper', () => {
    describe('actionToState', () => {
        it('should provide initial form state from scratch', () => {
            const newContactFieldFormState = formHelper.actionToState(
                null,
                Types.set_contact_field
            );

            expect(newContactFieldFormState.value).toEqual({ value: '' });
            expect(newContactFieldFormState).toMatchSnapshot(Types.set_contact_field);

            const newContactNameFormState = formHelper.actionToState(null, Types.set_contact_name);

            expect(newContactNameFormState.value).toEqual({ value: '' });
            expect(newContactNameFormState).toMatchSnapshot(Types.set_contact_name);
        });
    });

    describe('stateToAction', () => {
        it('should convert form state to an action', () => {
            expect(
                formHelper.stateToAction(
                    setContactFieldAction.uuid,
                    setContactFieldFormState,
                    Types.set_contact_field
                )
            ).toEqual(setContactFieldAction);

            expect(
                formHelper.stateToAction(
                    setContactNameAction.uuid,
                    setContactNameFormState,
                    Types.set_contact_name
                )
            ).toEqual(setContactNameAction);
        });
    });
});
