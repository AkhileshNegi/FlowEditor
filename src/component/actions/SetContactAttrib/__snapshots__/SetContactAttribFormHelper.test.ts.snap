// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetContactAttribFormHelper actionToState should provide initial form state from scratch: set_contact_field 1`] = `
Object {
  "field": Object {
    "id": "",
    "name": "",
    "type": "field",
  },
  "value": "",
}
`;

exports[`SetContactAttribFormHelper actionToState should provide initial form state from scratch: set_contact_name 1`] = `
Object {
  "name": Object {
    "id": "name",
    "name": "Name",
    "type": "name",
  },
  "value": "",
}
`;
